python_modules/dagster/dagster/_core/definitions/unresolved_asset_sensor_definition.py

```diff
@@ -0,0 +1,266 @@
+        update the cursor
+        """
+
+        def sensor_fn(context):
+            cursor_dict = json.loads(context.cursor) if context.cursor else {}
+            print(f"CURSOR {cursor_dict}")
+            should_materialize = []
+            cursor_update_dict = {}
+            for a in upstream.keys():
+                a_cursor = cursor_dict.get(str(a)) if cursor_dict.get(str(a)) else {}
+                (
+                    parent_asset_event_records,
+                    a_cursor_update_dict,
+                ) = self._get_parent_materializations(context, a, upstream[a], a_cursor)
+
+                if all(
+                    parent_asset_event_records.values()
+                ):  # should we allow the user to specify all() vs any() here?
+                    should_materialize.append(a)
+                    cursor_update_dict[str(a)] = a_cursor_update_dict
+
+            print(f"SHOULD MATERIALIZE {should_materialize}")
+
+            if len(should_materialize) > 0:
+                context.update_cursor(json.dumps(cursor_update_dict))
+                return RunRequest(run_key=f"{context.cursor}", asset_selection=should_materialize)
+
+        return SensorDefinition(
+            evaluation_fn=sensor_fn,
+            name=self.name,
+            job_name="__ASSET_JOB",
+            minimum_interval_seconds=self._minimum_interval_seconds,
+        )
+
+    def resolve(
+        self,
+        assets: Sequence["AssetsDefinition"],
+        source_assets: Sequence["SourceAsset"],
+    ):
+        return self._make_sensor(self._get_upstream_mapping(assets, source_assets))
+
+
+# class UnresolvedAssetSensorDefinitionV2(UnresolvedAssetSensorDefinition):
+#     def __init__(self, asset_selection, name):
+#         self._selection = asset_selection
+#         self._name = name
+
+#         super(UnresolvedAssetSensorDefinitionV2, self).__init__()
+
+#     @property
+#     def name(self):
+#         return self._name
+
+#     def get_dependency_mapping(
+#         self,
+#         assets: Sequence["AssetsDefinition"],
+#         source_assets: Sequence["SourceAsset"],
+#     ):
+#         """Computes a mapping of assets in self._selection to their parents in the asset graph"""
+#         upstream = defaultdict(set)
+#         downstream = defaultdict(set)
+#         selection_resolved = list(self._selection.resolve([*assets, *source_assets]))
+#         for a in selection_resolved:
+#             a_parents = list(AssetSelection.keys(a).upstream().resolve([*assets, *source_assets]))
+#             # filter out a because upstream() includes the assets in the original AssetSelection
+#             upstream[a] = {p for p in a_parents if p != a}
+#             for p in upstream[a]:
+#                 downstream[p].add(a)
+#         return {"upstream": upstream, "downstream": downstream}
+
+#     def make_sensor(self, dependency_mapping: Mapping[AssetKey, Sequence[AssetKey]]):
```

Comment:
alternate implementation that uses a combo of cursor and materialization time to determine if an asset should be materialized